reflectionserializer/manager.go

```diff
@@ -0,0 +1,795 @@
+func (m *SerializationManager) SerializeValue(value reflect.Value, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) error {
+	var err error
+	switch value.Kind() {
+	case reflect.Bool:
+		buffer.WriteBool(value.Bool())
+	case reflect.Int8:
+		buffer.WriteInt8(int8(value.Int()))
+	case reflect.Int16:
+		buffer.WriteInt16(int16(value.Int()))
+	case reflect.Int32:
+		buffer.WriteInt32(int32(value.Int()))
+	case reflect.Int64, reflect.Int:
+		buffer.WriteInt64(value.Int())
+	case reflect.Uint8:
+		buffer.WriteUint8(uint8(value.Uint()))
+	case reflect.Uint16:
+		buffer.WriteUint16(uint16(value.Uint()))
+	case reflect.Uint32:
+		buffer.WriteUint32(uint32(value.Uint()))
+	case reflect.Uint64, reflect.Uint:
+		buffer.WriteUint64(value.Uint())
+	case reflect.Float32:
+		if math.IsNaN(value.Float()) {
+			return fmt.Errorf("NaN float value")
+		}
+		buffer.WriteFloat32(float32(value.Float()))
+	case reflect.Float64:
+		if math.IsNaN(value.Float()) {
+			return fmt.Errorf("NaN float value")
+		}
+		buffer.WriteFloat64(value.Float())
+	case reflect.String:
+		err = WriteLen(len(value.String()), fieldMetadata.LengthPrefixType, buffer)
+		if err == nil {
+			buffer.WriteBytes([]byte(value.String()))
+		}
+	case reflect.Array:
+		err = m.serializeArray(value, fieldMetadata, buffer)
+	case reflect.Slice:
+		err = m.serializeSlice(value, fieldMetadata, buffer)
+	case reflect.Map:
+		err = fmt.Errorf("native map type is not supported. use orderedmap instead")
+	case reflect.Ptr:
+		err = m.serializePointer(value, fieldMetadata, buffer)
+	case reflect.Struct:
+		err = m.serializeStructure(value, fieldMetadata, buffer)
+	case reflect.Interface:
+		err = m.serializeInterface(value, fieldMetadata, buffer)
+	}
+	return err
+}
+
+func (m *SerializationManager) serializeInterface(value reflect.Value, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) error {
+	// write first byte only if AllowNil set to true
+	if value.IsNil() && fieldMetadata.AllowNil {
+		buffer.WriteByte(byte(0))
+		return nil
+	} else if fieldMetadata.AllowNil {
+		buffer.WriteByte(byte(1))
+	}
+
+	interfaceType := reflect.TypeOf(value.Interface())
+	encodedType, err := m.EncodeType(interfaceType)
+	if err != nil {
+		return err
+	}
+	buffer.WriteUint32(encodedType)
+	err = m.SerializeValue(value.Elem(), fieldMetadata, buffer)
+	return err
+}
+
+func (m *SerializationManager) serializeStructure(value reflect.Value, fieldMetadata FieldMetadata, buffer *marshalutil.MarshalUtil) error {
+	valueType := value.Type()
+	// individual serialization for time.Time
+	if valueType == reflect.TypeOf(time.Time{}) {
+		buffer.WriteTime(value.Interface().(time.Time))
+		return nil
+	} else if processed, err := m.tryBuiltInSerializer(value, valueType, fieldMetadata, buffer); err != nil || processed {
+		// serialize using built-in serializer if it's available
+		return err
+	} else {
+		// serialize as regular struct
+		err = m.serializeFields(value, buffer)
+		return err
+	}
```

Comment:
Use guard programming, i.e don't use if/else/else if, just early return (which you're doing anyway)